File: pd_ros/src/plutodrone/src/cam_pub.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <image_transport/image_transport.h>

int main(int argc, char** argv)
{
    // Initialize the ROS node
    ros::init(argc, argv, "camera_feed_publisher");
    ros::NodeHandle nh;

    // Create an image transport publisher to publish to the /camera/image_raw topic
    image_transport::ImageTransport it(nh);
    image_transport::Publisher image_pub = it.advertise("/camera/image_raw", 10);

    // Open the default camera (usually the webcam)
    cv::VideoCapture cap(6);

    // Check if the camera opened successfully
    if (!cap.isOpened())
    {
        ROS_ERROR("Could not open the camera!");
        return -1;
    }

    // Set the publishing rate (10 Hz)
    ros::Rate loop_rate(10);

    while (ros::ok())
    {
        // Capture a frame from the camera
        cv::Mat frame;
        cap >> frame;

        if (!frame.empty())
        {
            // Convert the captured frame (cv::Mat) to a ROS Image message
            cv_bridge::CvImagePtr cv_ptr;
            try
            {
                cv_ptr = cv_bridge::CvImagePtr(new cv_bridge::CvImage);
                cv_ptr->header.stamp = ros::Time::now(); // Time stamp
                cv_ptr->header.frame_id = "camera";      // Frame ID
                cv_ptr->encoding = sensor_msgs::image_encodings::BGR8; // Encoding format
                cv_ptr->image = frame; // The captured frame

                // Publish the image
                image_pub.publish(cv_ptr->toImageMsg());
                ROS_INFO("publishing");
            }
            catch (cv_bridge::Exception& e)
            {
                ROS_ERROR("cv_bridge exception: %s", e.what());
            }
        }

        // Spin once to handle callbacks and sleep to maintain the loop rate
        ros::spinOnce();
        loop_rate.sleep();
    }

    // Release the camera when the node is shutdown
    cap.release();

    return 0;
}

File: pd_ros/src/plutodrone/src/cam_pub.py
#!/usr/bin/env python

import rospy
import cv2
from cv_bridge import CvBridge
from sensor_msgs.msg import Image

def camera_feed_publisher():
    # Initialize the ROS node
    rospy.init_node('camera_feed_publisher', anonymous=True)

    # Set up the publisher to publish to the topic "/camera/image_raw"
    image_pub = rospy.Publisher('/camera/image_raw', Image, queue_size=10)

    # Create a CvBridge object to convert OpenCV images to ROS Image messages
    bridge = CvBridge()

    # Open a connection to the camera (0 is usually the default webcam)
    cap = cv2.VideoCapture(6)

    # Check if the camera is opened properly
    if not cap.isOpened():
        rospy.logerr("Could not open camera.")
        return

    rate = rospy.Rate(10)  # Set the publish rate to 10 Hz

    while not rospy.is_shutdown():
        # Capture frame-by-frame
        ret, frame = cap.read()

        if ret:
            # Convert the OpenCV image (BGR format) to a ROS Image message
            ros_image = bridge.cv2_to_imgmsg(frame, encoding="bgr8")

            # Publish the image
            image_pub.publish(ros_image)

            # Log the publishing status
            rospy.loginfo("Published a new frame to /camera/image_raw")
        
        # Sleep to maintain the desired publishing rate
        rate.sleep()

    # Release the camera when the node is shut down
    cap.release()

if __name__ == '__main__':
    try:
        camera_feed_publisher()
    except rospy.ROSInterruptException:
        pass
